packages/appcache-polyfill-sw/index.ts

```diff
@@ -0,0 +1,318 @@
+  const entries = Object.entries(clientIdToHash);
+  for (const [clientId, hash] of entries) {
+    if (idsOfActiveClients.indexOf(clientId) === -1) {
+      delete clientIdToHash[clientId];
+      inactiveHashes.push(hash);
+    }
+  }
+
+  await storage.set('ClientIdToHash', clientIdToHash);
+
+  return inactiveHashes;
+}
+
+async function getHashesOfOlderVersions() {
+  const hashesOfOlderVersions: Set<string> = new Set();
+
+  const manifestURLToHashes: ManifestURLToHashes =
+      await storage.get('ManifestURLToHashes');
+
+  for (const hashToManifest of Object.values(manifestURLToHashes)) {
+    const allHashes = Array.from(hashToManifest.keys());
+    // We want to iterate over everything other than the last key.
+    for (const hash of allHashes.slice(0, allHashes.length - 1)) {
+      hashesOfOlderVersions.add(hash);
+    }
+  }
+
+  return hashesOfOlderVersions;
+}
+
+async function removeUnusedHashAssociations(hashes: Array<string>) {
+  const manifestURLToHashes: ManifestURLToHashes =
+      await storage.get('ManifestURLToHashes');
+
+  for (const hashToManifest of Object.values(manifestURLToHashes)) {
+    for (const hash of Object.keys(hashToManifest)) {
+      if (hashes.includes(hash)) {
+        hashToManifest.delete(hash);
+      }
+    }
+  }
+
+  await storage.set('ManifestURLToHashes', manifestURLToHashes);
+}
+
+/**
+ * Does the following:
+ * 1. Gets a list of all client ids associated with this service worker.
+ * 2. Calls cleanupClientIdAndHash() to remove the out of date client id
+ *    to hash associations.
+ * 3. Calls getHashesOfOlderVersions() to get a list of all the hashes
+ *    that correspond to out-of-date manifest versions.
+ * 4. If there's a match between an out of date hash and a hash that is no
+ *    longer being used by a client, then it deletes the corresponding cache.
+ */
+async function cleanup() {
+  const activeClients = await self.clients.matchAll();
+  const idsOfActiveClients = activeClients.map((client) => client.id);
+  const hashesNotInUse = await cleanupClientIdAndHash(idsOfActiveClients);
+  const hashesOfOlderVersions = await getHashesOfOlderVersions();
+  const hashesToDelete = Array.from(hashesOfOlderVersions).filter(
+      (hash) => hashesNotInUse.includes(hash));
+
+  // Now that we know what hashes are no longer used, cleanup two things:
+  // Delete the IndexedDB associations:
+  await removeUnusedHashAssociations(hashesToDelete);
+  // Delete the Cache Storage entries:
+  await Promise.all(hashesToDelete.map((hash) => caches.delete(hash)));
+}
+
+export async function handle(event: FetchEvent) {
+  const response = await appCacheBehaviorForEvent(event);
+  // If this is a navigation, clean up unused caches that correspond to old
+  // AppCache manifest versions which are no longer associated with an
+  // active client. This will be done asynchronously, and won't block the
+  // response from being returned to the fetch handler.
+  if (event.request.mode === 'navigate') {
```

Comment:
Alternatively, I could trigger a cleanup each time the SW starts up, but this seemed sufficient.